### PR TITLE
Implement for_each and each.* completion, hover info, and semantic tokens

### PR DESCRIPTION
**Pull request by jpogran**
This provides completion hints inside blocks for `for_each` and `each.*` references within `resource`, `data` and `module` blocks anywhere the `for_each` meta-argument is supported. It detects if `for_each` is used already and does not suggest duplicates and detects when it is necessary to provide `each.*` completions. This does not complete the values for `each.key` and `each.value`, e.g. `each.value.something`.

This also provides hover support for `for_each` and `each.*` references using the documentation as source.

This also provides semantic token support for `for_each` and `each.*` references.

Closes https://github.com/hashicorp/terraform-ls/issues/861

### CODE REVIEW

File: decoder/decoder.go [hunk 1]
```diff
@@ -172,6 +175,17 @@ func countAttributeSchema() *schema.AttributeSchema {
 	}
 }
 
+func forEachAttributeSchema() *schema.AttributeSchema {
+	return &schema.AttributeSchema{
+		IsOptional: true,
+		Expr: schema.ExprConstraints{
+			schema.TraversalExpr{OfType: cty.Map(cty.String)},
+			schema.LiteralTypeExpr{Type: cty.Set(cty.String)},
```
Comment (jpogran) [inline decoder/decoder.go:183]: Should this match https://github.com/hashicorp/terraform-schema/blob/aac2bff86a0b7014a2d2b70044e6558b2edc3fd6/internal/schema/0.12/resource_block.go#L80-L91?

File: decoder/expression_candidates.go [hunk 2]
```diff
@@ -331,6 +331,10 @@ func (d *PathDecoder) constraintToCandidates(ctx context.Context, constraint sch
 		if schema.ActiveCountFromContext(ctx) && attr.Name != "count" {
 			candidates = append(candidates, countIndexCandidate(editRng))
 		}
+		if schema.ActiveForEachFromContext(ctx) {
```
Comment (radeksimko) [inline decoder/expression_candidates.go:334]: Since it also doesn't make sense to complete `each.*` within `for_each = HERE`, similar to how we don't complete `count.index` in `count = HERE` I think we'll need some check for that + ideally a test too.

File: decoder/semantic_tokens.go [hunk 3]
```diff
@@ -606,3 +601,33 @@ func tokensForTupleConsExpr(expr *hclsyntax.TupleConsExpr, exprType cty.Type) []
 
 	return tokens
 }
+
+func tokensForTraversalExpression(traversal hcl.Traversal) []lang.SemanticToken {
```
Comment (radeksimko) [inline decoder/semantic_tokens.go:605]: Nitpick: Since we're operating in the `decoder` package namespace, I think we should make this a little more specific, e.g. `semanticTokensForTraversalExpression`?

File: decoder/body_extensions_test.go [hunk 4]
```diff
@@ -460,6 +460,452 @@ variable "test" {
...
+								Line:   2,
+								Column: 14,
+								Byte:   45,
+							},
+						},
+						NewText: "each.value",
+						Snippet: "each.value",
+					},
+					Kind: lang.TraversalCandidateKind,
+				},
```
Comment (radeksimko) [inline decoder/body_extensions_test.go:571]:
```suggestion
```
As mentioned in my other comment in-line, these two references shouldn't be available in the context of `for_each` directly, that would be kinda like a reference to itself.

File: decoder/body_extensions_test.go [hunk 5]
```diff
@@ -460,6 +460,452 @@ variable "test" {
...
+						},
+						NewText: "type",
+						Snippet: `type = "${1:value}"`,
+					},
+					Kind: lang.AttributeCandidateKind,
+				},
+			}),
+		},
+		{
+			"for_each does not complete with count present",
```
Comment (radeksimko) [inline decoder/body_extensions_test.go:719]: This is a useful test, but it would be also nice to validate it earlier, before the [invalid] schema even gets to the decoder, here: https://github.com/hashicorp/hcl-lang/blob/e12fb9092dc7bde532c1b1346069a23d9eb2c7f9/schema/body_schema.go#L154-L181

This validation provides some form of basic "integrity checks" and it is something we typically run as part of tests within `terraform-schema`.

It doesn't negate the need to catch it here also, but it can be an additional early warning.

**Review by radeksimko — COMMENTED** [review]
This is looking pretty good. Aside from my in-line comments we're also missing hover tests.

On the note of testing - in general, when I'm unsure what needs testing, I'd use the coverage report which is built into Go:

```
go test -coverprofile=coverage.out ./...
```
```
go tool cover -html=coverage.out
```
https://go.dev/blog/cover

That's not to say I'd expect you to cover 100%, but I think it can provide a good guidance.

File: decoder/candidates.go [hunk 6]
```diff
@@ -55,13 +55,23 @@ func (d *PathDecoder) candidatesAtPos(ctx context.Context, body *hclsyntax.Body,
 				ctx = schema.WithActiveCount(ctx)
 			}
 		}
+
+		if bodySchema.Extensions.ForEach {
+			if _, ok := body.Attributes["count"]; !ok {
```
[the remark this answers is not in the document]
Comment (jpogran) [reply decoder/candidates.go:60]: In the docs it says for_each and count can't be used in the same block, so as with body_candidates, I thought it needed to be checked against count too here.

File: decoder/decoder.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpogran) [reply decoder/decoder.go:183]: Thanks for confirming my question above

File: decoder/decoder.go [hunk 7]
```diff
@@ -192,3 +206,90 @@ func countIndexCandidate(editRng hcl.Range) lang.Candidate {
...
+				"**Note**: A given resource or module block cannot use both count and for_each.",
+		},
+		Kind: lang.AttributeCandidateKind,
+		TextEdit: lang.TextEdit{
+			NewText: "for_each",
+			Snippet: "for_each {\n ${1}\n}",
+			Range:   editRng,
+		},
+	}
+}
```
[the remark this answers is not in the document]
Comment (jpogran) [reply decoder/decoder.go:228]: Correct, was leftover from the transition to the changes in previous PR

File: decoder/semantic_tokens.go [hunk 8]
```diff
@@ -164,42 +172,29 @@ func (d *PathDecoder) tokensForExpression(ctx context.Context, expr hclsyntax.Ex
...
+
 		countAvailable := schema.ActiveCountFromContext(ctx)
 		countIndexAttr := lang.Address{
-			lang.RootStep{
-				Name: "count",
-			},
-			lang.AttrStep{
-				Name: "index",
-			},
+			lang.RootStep{Name: "count"}, lang.AttrStep{Name: "index"},
```
[the remark this answers is not in the document]
Comment (jpogran) [reply decoder/semantic_tokens.go:178]: No problem. In adding each.*, I wondered if these addresses should be extracted out of this function, or if this code should be a switch statement instead of an if/else since we expect to add more (dynamic, etc).

File: decoder/semantic_tokens.go (hunk 3 above)
Comment (jpogran) [reply decoder/semantic_tokens.go:605]: Not a nitpick, its defensive programming 😁

File: decoder/semantic_tokens.go [hunk 9]
```diff
@@ -606,3 +601,33 @@ func tokensForTupleConsExpr(expr *hclsyntax.TupleConsExpr, exprType cty.Type) []
...
+				Column: traversal[1].SourceRange().Start.Column + 1,
+				Byte:   traversal[1].SourceRange().Start.Byte + 1,
+			},
+			End: hcl.Pos{
+				Line:   traversal[1].SourceRange().End.Line,
+				Column: traversal[1].SourceRange().End.Column + 1,
+				Byte:   traversal[1].SourceRange().End.Byte + 1,
+			},
+		},
+	})
```
[the remark this answers is not in the document]
Comment (jpogran) [reply decoder/semantic_tokens.go:630]: Good point, defensive programming is always good 😁

File: decoder/candidates.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (radeksimko) [reply decoder/candidates.go:60]: Ah, yes - sorry - I misread the meaning of this. Maybe adding a comment on what the condition does would be useful? 😅

File: decoder/decoder.go [hunk 10]
```diff
@@ -172,6 +175,17 @@ func countAttributeSchema() *schema.AttributeSchema {
...
 }
 
+func forEachAttributeSchema() *schema.AttributeSchema {
+	return &schema.AttributeSchema{
+		IsOptional: true,
+		Expr: schema.ExprConstraints{
+			schema.TraversalExpr{OfType: cty.Map(cty.String)},
+			schema.LiteralTypeExpr{Type: cty.Set(cty.String)},
+		},
+		Description: lang.Markdown("A meta-argument that accepts a map or a set of strings, "),
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply decoder/decoder.go]: I think that is basically what we're doing by declaring the `schema.*` types and reading it from there at runtime in the different completion/hover/semtok methods?

File: decoder/semantic_tokens.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (radeksimko) [reply decoder/semantic_tokens.go:178]: It's something I plan to address as part of go-to-definition/go-to-references - I'm hoping that once we treat `each.*` as actual reference {target,origin} then we can just leverage the existing reference matching logic and this address here shouldn't be necessary as we compare it against _all_ valid references instead.

TL;DR I wouldn't worry about it here.

File: decoder/candidates.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jpogran) [reply decoder/candidates.go]: After discussion, we've decided to not check for the existence of count of foreach and not include it in the completion results if the other is present. We believe completion is not the place for validation

File: decoder/hover_test.go [hunk 11]
```diff
@@ -1073,3 +1073,168 @@ func TestDecoder_HoverAtPos_extension(t *testing.T) {
...
+				Content: lang.Markdown("`count.index` _number_\n\nThe distinct index number (starting with 0) corresponding to the instance"),
+				Range: hcl.Range{
+					Filename: "test.tf",
+					Start:    hcl.Pos{Line: 3, Column: 11, Byte: 44},
+					End:      hcl.Pos{Line: 3, Column: 22, Byte: 55},
+				},
+			},
+		},
+		{
+			"count attribute value",
```
Comment (radeksimko) [inline decoder/hover_test.go]:
```suggestion
			"for_each attribute value",
```

File: decoder/body_extensions_test.go [hunk 12]
```diff
@@ -380,83 +380,404 @@ func TestCompletionAtPos_BodySchema_Extensions(t *testing.T) {
 			}),
 		},
 		{
-			"count attribute expression completion",
+			"foreach attribute completion",
```
Comment (radeksimko) [inline decoder/body_extensions_test.go:383]: Unless I'm misreading the diff, I think some tests are being removed here?

**Review by radeksimko — APPROVED** [review]
Just one question left about the test, otherwise LGTM.

If you could also squash the commits, that would be great.
